pupepat/test/pupil-synth/test/test_pupil_plate_image.py

```diff
@@ -0,0 +1,179 @@
+
+        # check that the Source now knows where it is on the plate
+        self.assertEqual((cx, cy), ppi.sources[0].position)
+
+        # check that the Source is actually there.
+        self.assertEqual(ps.counts, image[ps.x, ps.y])
+
+    def test_pixel_perfect_placement(self):
+        ppi = PupilPlateImage((4, 4))  # typical even number of pixels/edge
+        ppi.add_poisson_noise = False
+        ppi.sky_level = 0
+        ppi.read_noise = 0
+
+        radius, counts = 1, 99  # radius 0 gives single pixel decal
+        ps = PointSource(radius, counts)
+
+        # add the single pixel point source to the plate
+        cx = 2
+        cy = 1
+        ppi.add_source(ps, cx, cy)
+
+        image = ppi.render()
+
+        # check that source is where we placed it
+        for x in range(image.shape[0]):
+            for y in range(image.shape[1]):
+                if not ((x == cx) and (y == cy)):
+                    self.assertEqual(0, image[x, y], '({x:}, {y:})'.format(x=x, y=y))
+        self.assertEqual(counts, image[cx, cy], '({x:}, {y:})'.format(x=cx, y=cy))
+
+
+class TestPupilPlateImageCutoutExtraction(PupilSynthTestCase):
+
+    def setUp(self):
+        self.ppi = PupilPlateImage(shape=(1024, 1024))
+        self.cp = CircularPupil(100, 200, 1000.)
+        self.ppi.add_source(self.cp, 512, 512)
+
+    def test_cutout_margin_factor(self):
+        expected_default = 1.2  # from the class definition
+        # check the default value
+        self.assertEqual(expected_default, self.ppi.cutout_margin_factor)
+
+        # check setting/getting
+        new_value = 1.618
+        self.ppi.cutout_margin_factor = new_value
+        self.assertEqual(new_value, self.ppi.cutout_margin_factor)
+
+    def test_cutout_for_source_cutout(self):
+        # if we set the cutout_margin_factor to 1.0,
+        self.ppi.cutout_margin_factor = 1.0
+        # then the shape of the decal and cutout should be equal
+        cutout = self.ppi.cutout_for_source(self.cp)
+        self.assertEqual(self.cp.decal.shape, cutout.shape)
+
+
+#@skip
+class TestPupilPlateImageFileIO(PupilSynthTestCase):
+    """It might seem like we're just testing astropy.io.fits here,
+    but I think we'll try to write the synthesized pupil parameters
+    into the header pretty soon. That prepares us to run a FITS file
+    through PUPE-PAT and know it's own correct answers.
+    """
+
+    def setUp(self):
+        self.ppi = PupilPlateImage(shape=(1024, 1024))
+        self.cp = CircularPupil(100, 200, 1000.)
+        self.ppi.add_source(self.cp, 512, 512)
+
+    def test_write_and_read(self):
+        # create a temporary file with a name
+        temp_file = tempfile.NamedTemporaryFile(dir='/tmp')
+        temp_file.close()
+
+        # write the PupilPlateImage to the file
+        self.ppi.write_fits_to(temp_file.name)
+        # read it into a new PPI instance
+        ppi_from_file = PupilPlateImage()
+        ppi_from_file.read_fits_from(temp_file.name)
```

Comment:
Do we really need the cutouts to write to disk? Couldn't we just test everything in memory?